src/lib/raft/raft.c

```diff
@@ -723,13 +723,17 @@ raft_worker_handle_io(struct raft *raft)
 					raft_sm_become_leader(raft);
 				else
 					raft_sm_become_candidate(raft);
-			} else if (raft->vote != 0) {
+			} else if (raft->vote != 0 || raft->last_demoted_term ==
+						      raft->volatile_term) {
 				/*
-				 * Voted for some other node. Wait if it manages
-				 * to become a leader.
+				 * Voted for some other node or resigned from
+				 * leadership in the current turn (i.e., cannot
+				 * become a candidate in the current elections),
+				 * so just wait.
 				 */
 				raft_sm_wait_election_end(raft);
-			} else {
+			} else if (raft->last_demoted_term <
```

Comment:
Nit: it seems safer to make "voting" a default path and just leave `else without condition. `last_demoted_term` should be in assert.

Reply:
Yeah, sorry I missed this after you suggest to merge the branches — fixed.